libraries/persistent_store/src/store.rs

```diff
@@ -128,3 +168,1089 @@ impl StoreUpdate {
+                ParsedEntry::Tail => break,
+                ParsedEntry::User(header) if sorted_keys.binary_search(&header.key).is_ok() => {
+                    self.delete_pos(entry_pos, pos - entry_pos - 1)?;
+                }
+                ParsedEntry::Padding | ParsedEntry::User(_) => (),
+                _ => return Err(StoreError::InvalidStorage),
+            }
+        }
+        Ok(())
+    }
+
+    /// Deletes the entry at a given position.
+    fn delete_pos(&mut self, pos: Position, length: Nat) -> StoreResult<()> {
+        self.set_deleted(pos)?;
+        self.wipe_span(pos + 1, length)?;
+        Ok(())
+    }
+
+    /// Writes the init info of a page between 2 positions if needed.
+    ///
+    /// The positions should designate the first and last word of an entry. The init info of the
+    /// highest page is written in any of the following conditions:
+    /// - The entry starts at the beginning of a virtual page.
+    /// - The entry spans 2 pages.
+    fn init_page(&mut self, first: Position, last: Position) -> StoreResult<()> {
+        debug_assert!(first <= last);
+        debug_assert!(last - first < self.format.virt_page_size());
+        let new_first = if first.word(&self.format) == 0 {
+            first
+        } else if first.page(&self.format) == last.page(&self.format) {
+            return Ok(());
+        } else {
+            last + 1
+        };
+        let page = new_first.page(&self.format);
+        if let WordState::Valid(_) = self.parse_init(page)? {
+            return Ok(());
+        }
+        let index = self.format.index_init(page);
+        let init_info = self.format.build_init(InitInfo {
+            cycle: new_first.cycle(&self.format),
+            prefix: new_first.word(&self.format),
+        });
+        self.storage_write_slice(index, &init_info)?;
+        Ok(())
+    }
+
+    /// Sets the padding bit of a user header.
+    fn set_padding(&mut self, pos: Position) -> StoreResult<()> {
+        let mut word = Word::from_slice(self.read_word(pos));
+        self.format.set_padding(&mut word);
+        self.write_slice(pos, &word.as_slice())?;
+        Ok(())
+    }
+
+    /// Sets the deleted bit of a user header.
+    fn set_deleted(&mut self, pos: Position) -> StoreResult<()> {
+        let mut word = Word::from_slice(self.read_word(pos));
+        self.format.set_deleted(&mut word);
+        self.write_slice(pos, &word.as_slice())?;
+        Ok(())
+    }
+
+    /// Wipes a slice of words.
+    fn wipe_span(&mut self, pos: Position, length: Nat) -> StoreResult<()> {
+        let length = (length * self.format.word_size()) as usize;
+        self.write_slice(pos, &vec![0x00; length])
+    }
+
+    /// Returns an extremum page.
+    ///
+    /// With `Greater` returns the most recent page (or the tail). With `Less` returns the oldest
+    /// page (or the head).
+    fn get_extremum_page_head(&self, ordering: Ordering) -> StoreResult<Position> {
```

Comment:
I'm wondering if you ever tested the storage implementation's speed vs. the old version. You are looping here, and this function gets called by `head()`, and `tail()`. So quite often, I think?

Reply:
I did not do a proper comparison, I just have a latency example application for each but they don't really do the same (since it depends on the actual implementation). However, I would say both behave similarly (because all operations are linear and the slowest things are writes, erases, and syscalls, which are done a similar number of times).

The loop here is short, it's in the number of pages. But yes, the first optimization we should do if we want to improve latency and lose a bit of memory would be to store `head` and `tail` in memory. That was planned but I prefer to do it later and only if needed since it's an optimization.

Reply:
The store wasn't really stress tested for lookup speed in v1 either, as far as I recall, so I guess that's fair.